[Error] Show pretty & intuitive stack traceback when ti.enable_excepthook() used (#1405)
taichi-dev/taichi: docs/debugging.rst

        return x % 2 == 1

Pretty Taichi-scope traceback
-----------------------------

As we all know, Python provides a useful stack traceback system, which could help you

> locale the issue easier. But sometimes stack tracebacks from **Taichi-scope** could be

Reviewer: ```suggestion
locate the issue easily. But sometimes stack tracebacks from **Taichi-scope** could be
```

        func1()
      File "/root/taichi/python/taichi/lang/kernel.py", line 39, in decorated
        return fun.__call__(*args)
      File "/root/taichi/python/taichi/lang/kernel.py", line 79, in __call__
        ret = self.compiled(*args)
      File "misc/demo_excepthook.py", line 14, in func1
        func2()
      File "/root/taichi/python/taichi/lang/kernel.py", line 39, in decorated
        return fun.__call__(*args)
      File "/root/taichi/python/taichi/lang/kernel.py", line 79, in __call__
        ret = self.compiled(*args)
      File "misc/demo_excepthook.py", line 10, in func2
        func3()
      File "/root/taichi/python/taichi/lang/kernel.py", line 39, in decorated
        return fun.__call__(*args)
      File "/root/taichi/python/taichi/lang/kernel.py", line 79, in __call__
        ret = self.compiled(*args)
      File "misc/demo_excepthook.py", line 6, in func3
        ti.static_assert(1 + 1 == 3)
      File "/root/taichi/python/taichi/lang/error.py", line 14, in wrapped
        return foo(*args, **kwargs)
      File "/root/taichi/python/taichi/lang/impl.py", line 252, in static_assert
        assert cond
    AssertionError

> You may already being brain fried by the annoying ``decorated``'s and ``__call__``'s.

Reviewer: ```suggestion
You may already feel brain fried by the annoying ``decorated``'s and ``__call__``'s.
```

> These are the Taichi internal stack frames, has almost no benefit for end-users but

Reviewer: ```suggestion
These are the Taichi internal stack frames. They have almost no benefit for end-users but
```

> making the traceback hard to read.

Reviewer: ```suggestion
make the traceback hard to read.
```

    @ti.kernel
    def func0():
        func1()

    func0()  <--
    --------------------------------------------
    In func0() at misc/demo_excepthook.py:19:
    --------------------------------------------
        func2()

    @ti.kernel
    def func0():
        func1()  <--

    func0()
    --------------------------------------------
    In func1() at misc/demo_excepthook.py:15:
    --------------------------------------------
        func3()

    @ti.func
    def func1():
        func2()  <--

    @ti.kernel
    --------------------------------------------
    In func2() at misc/demo_excepthook.py:11:
    --------------------------------------------
        ti.static_assert(1 + 1 == 3)

    @ti.func
    def func2():
        func3()  <--

    @ti.func
    --------------------------------------------
    In func3() at misc/demo_excepthook.py:7:
    --------------------------------------------
    ti.enable_excepthook()

    @ti.func
    def func3():
        ti.static_assert(1 + 1 == 3)  <--

    @ti.func
    --------------------------------------------
    AssertionError

See? Our exception hook has removed some useless Taichi internal frames from

> traceback, and what's more, although not visible in doc, the output is

Reviewer: ```suggestion
traceback. What's more, although not visible in the doc, the output is
```
